library/public/types.h

```diff
@@ -82,6 +72,103 @@ struct mesh_t
    */
   F3D_EXPORT std::pair<bool, std::string> isValid() const;
 };
+
+/**
+ * Describe a 3D vector.
+ */
+struct vector3_t
+{
+  vector3_t() = default;
+  vector3_t(double x, double y, double z)
+    : Value{ x, y, z }
+  {
+  }
+  vector3_t(const std::vector<double>& vec)
+  {
+    if (vec.size() != 3)
+    {
+      throw std::runtime_error("cannot create a vector3_t");
+    }
+    Value[0] = vec[0];
+    Value[1] = vec[1];
+    Value[2] = vec[2];
+  }
+
+  double* data()
+  {
+    return Value;
+  }
+  const double* data() const
+  {
+    return Value;
+  }
+
+  double& operator[](int idx)
+  {
+    return Value[idx];
+  }
+  double operator[](int idx) const
+  {
+    return Value[idx];
+  }
+  operator std::vector<double>() const
+  {
+    return { Value[0], Value[1], Value[2] };
+  }
+  operator std::array<double, 3>() const
+  {
+    return { Value[0], Value[1], Value[2] };
+  }
+  bool operator==(const vector3_t& vec) const
+  {
+    return Value[0] == vec.Value[0] && Value[1] == vec.Value[1] && Value[2] == vec.Value[2];
+  }
+  bool operator!=(const vector3_t& vec) const
+  {
+    return !(*this == vec);
+  }
+
+  double* begin()
+  {
+    return Value;
+  }
+  const double* cbegin() const
+  {
+    return Value;
+  }
+  double* end()
+  {
+    return Value + 3;
+  }
+  const double* cend() const
+  {
+    return Value + 3;
+  }
+
+  static vector3_t fromSphericalCoordinates(double theta, double phi)
+  {
+    return { std::sin(phi) * std::cos(theta), std::sin(phi) * std::cos(theta), std::cos(phi) };
```

Comment:
Maybe cache sin(phi) and cos(theta) to a local variable since they are used twice.